src/interp/mod.rs

```diff
@@ -212,6 +218,35 @@ impl LLVMIRInterpreter {
                         Name(n) => {
                             if n.as_str() == "printf" {
                                 self.printf(call);
+                            } else {
+                                match self.module.get_func_by_name(n) {
+                                    Some(func) => {
+                                        self.func_vars_stack.push(self.vars.clone());
```

Comment:
We can do without cloning `vars` here I think:
```
self.func_vars_stack.push(self.vars);
self.vars = HashMap::new();
```

Reply:
Doesn't seem to work as 'cannot move out of self.vars which is behind a mutable reference.

Reply:
I see. I still don't feel good about cloning vars here when it's not strictly necessary. There are other ways to work around this, e.g. by not having `self.vars` at all, and make inserting a new variable always insert it into the map on top of the call stack (see `Vec::last_mut`). I'd probably refactor var inserting into its own function to do that.

Reply:
Yes, that might also work.

Reply:
Rc?

Reply:
https://doc.rust-lang.org/rust-by-example/std/rc.html

Reply:
https://doc.rust-lang.org/std/rc/struct.Rc.html -- it allows you to share things, although mutating them is difficult. It's probably not worth fixing in this PR, but worth looking at for the next. You might also want to ask @jacob-hughes about his `Gc` type, which is more suitable for interpreters (but does mean you can't use normal rustc at the moment).

Reply:
> It's probably not worth fixing in this PR

Should we fix the unnecessary clone in the meantime though?

Reply:
If the clone is unnecessary, we should definitely kill it! [`cargo clippy` might also be worth running at some point soon -- it often spots unnecessary `clone`s.]